Library/Homebrew/cmd/update-report.rb

```diff
@@ -101,6 +104,32 @@ def install_core_tap_if_necessary
     ENV["HOMEBREW_UPDATE_BEFORE_HOMEBREW_HOMEBREW_CORE"] = revision
     ENV["HOMEBREW_UPDATE_AFTER_HOMEBREW_HOMEBREW_CORE"] = revision
   end
+
+  def migrate_legacy_cache_if_necessary
```

Comment:
>On every file or just directories? I think it's worth moving some stuff, even if some files have bad permissions, no?

Yeah, on everything. The goal would be to have them fix it first. The permissions-fixer-tool-thingy could potentially be used to fix the permissions such that it would then pass the test. You can use `find /Library/Caches/Homebrew -not -uid ... -not -gid ...` etc.` to do the recursive check, which should turn up no results if it's good to go.

For same volume check, you can use the same method xucheng does for the size computation by comparing stat.dev: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/pathname.rb#L40

For checking whether migration has been attempted, probably a dot file at like `$(brew --prefix)/.sh.brew.didAttemptCacheMigration` as Apple does for things like recording whether the user said a given volume should never be presented as a possible Time Machine destination.